### PR TITLE
Optimise the queries powering the org dashboard page

### PR DESCRIPTION
**Pull request by samuelhwilliams**
## Summary
The org dashboard view can be really slow for some orgs with lots of services and notifications. The queries that power this view haven't been optimised much, and so we can make 100s or 1000s of round trips to the DB. Let's refactor some of the supporting logic to make queries in bulk so that we have fewer round trips, and hopefully (much) faster page loads.

With the test case provided below, on my machine we go from ~1250 DB queries to 11, and from 8.6s to 0.3s

This will also speed up the nightly billing fact generation task (create-nightly-billing-for-day). Over the last 7 days in prod it's taken approx 300s each time.

On my machine, with the local test data from below, on `main` it takes 18 seconds and on this branch it takes <1 second. (the log messages don't have any decimal precision, so it comes back as 0 seconds).

## Testing locally
I've added a script to generate an organisation with 100 services and 5,000,000 notifications across them, so that the org dashboard view loads slowly locally.

1) Run: `flask command generate-bulktest-data -u <YOUR_LOCAL_USER_ID>`. This runs pretty slowly - expect to kick it off and walk away for 30 minutes.

    <details><summary>It takes a while...</summary>

    ```
    ~/work/gds/notifications-api [SW-reduce-org-dashboard-queries]$ flask command generate-bulktest-data -u 0c38e85c-b037-4dd8-ae12-09ab2b84c373
    [0.00]: Sending org to DB...
    [0.05]: Done.
    [0.05]: Building services...
    [0.05]: Sending services to DB...
    [0.09]: Done.
    [0.09]: Building templates...
    [0.11]: Sending templates to DB...
    [0.11]: Done.
    [0.11]: Building 5,000,000 notifications...
    [0.18]: Building batch 1...
    [28.20]: Sending batch 1 to DB...
    [370.70]: Done.
    [370.70]: Building batch 2...
    [399.35]: Sending batch 2 to DB...
    
    [920.54]: Done.
    [920.54]: Building batch 3...
    [948.87]: Sending batch 3 to DB...
    [1578.88]: Done.
    [1578.88]: Building batch 4...
    [1607.78]: Sending batch 4 to DB...
    [2292.68]: Done.
    [2292.68]: Building batch 5...
    [2320.96]: Sending batch 5 to DB...
    [3081.49]: Done.
    [3081.49]: Committing...
    [3081.53]: Finished.
    ```
    </details>

    <details><summary>To delete all(?) the data afterwards</summary>

    ```
    delete from notifications where client_reference LIKE 'BULKTEST%';
    delete from notification_history where client_reference LIKE 'BULKTEST%';
    delete from template_redacted where "template_id" IN (select "id" from templates where "name" like 'BULKTEST%');
    delete from templates where "name" like 'BULKTEST%';
    delete from templates_history where "name" like 'BULKTEST%';
    delete from service_permissions where service_id IN (select "id" from services where "name" like 'BULKTEST%');
    delete from permissions where service_id IN (select "id" from services where "name" like 'BULKTEST%');
    delete from user_to_service where service_id in  (select "id" from services where "name" like 'BULKTEST%');
    delete from service_sms_senders where service_id in  (select "id" from services where "name" like 'BULKTEST%');
    delete from annual_billing where service_id in  (select "id" from services where "name" like 'BULKTEST%');
    delete from services where "name" like 'BULKTEST%';
    delete from organisation where "name" like 'BULKTEST%';
    ```
    </details>
2) I've pushed up a commit 7c9a6aeb17feecef9fced26a10a7b7d5603a7f0b on a separate branch that adds a profiler and flask-debugtoolbar, as well as a fake view that replicates the org services-with-usage endpoint used by the org view dashboard.
    `git checkout main`
    `git cherry-pick 7c9a6aeb17feecef9fced26a10a7b7d5603a7f0b`
    `make bootstrap`
3) Open your browser and navigate to http://localhost:6011/organisations/BULKTEST_ORG_ID_HERE/services-with-usage-html?year=2022. Check the toolbar on the side and record the time taken and the number of queries executed.
4) Put your `main` branch back to normal: `git reset --hard main` / `git reset --hard head^` to remove the cherry-picked commit
5) Checkout this branch, `git checkout SW-reduce-org-dashboard-queries` and cherry-pick the profiling commit again:  `git cherry-pick 7c9a6aeb17feecef9fced26a10a7b7d5603a7f0b`
6) Run step 3 again. Compare the time taken and queries executed.

Note: the numbers on the dashboard will feel 'wrong', because they are - we haven't built historical billing data. This could be done by running `create_nightly_billing_for_day(date)` for all dates in the range that the command generates notifications within. However this doesn't, as far as I can see, affect the performance of the dashboard - just the accuracy of the stats.

## Comparison

| Before | After |
|---|---|
| image | image |

### CODE REVIEW

File: tests/app/organisation/test_rest.py [hunk 1]
```diff
@@ -840,6 +841,30 @@ def test_get_organisation_services_usage(admin_request, notify_db_session):
...
+@pytest.mark.parametrize("num_services", [1, 5, 10])
+@freeze_time("2020-02-24 13:30")
+def test_get_organisation_services_usage_limit_queries_executed(admin_request, notify_db_session, num_services):
+    org = create_organisation(name="Organisation without live services")
+    for _ in range(num_services):
+        service = create_service(service_name=f"service {_}")
+        template = create_template(service=service)
+        dao_add_service_to_organisation(service=service, organisation_id=org.id)
+        create_annual_billing(service_id=service.id, free_sms_fragment_limit=10, financial_year_start=2019)
+        create_ft_billing(
```
Comment (idavidmcdonald) [inline tests/app/organisation/test_rest.py:853]: Should we be creating more rows of billing data for each service to make it more representative of production?

**Review by idavidmcdonald — COMMENTED** [review]
So I've done half a review on this as it is complex and it is now 6pm but I did want to try and move this forward.

I've gone through commit by commit reviewing the code.

I couldn't easily get my head around the second commit and it got too late for me to start doing it. It probably isn't that crazy but definitely touches some of our more complex code and I would need to do a fair bit of sqlalchemy docs reading. If you want to move this pull request through quicker, I'd be open to moving the second commit to a separate PR or dropping it (assuming that it is not the deal breaker commit that really reduces the bulk of the time). Such complex code of ours is hard to review and I'm not sure that your changes in that commit make it any easier (not that I would necessarily expect them to but if the performance benefits aren't drastic for the commit, I'm OK to not include it for that reason).

I also haven't reviewed the commit about the command to generate bulk data to save me time as it is a dev only thing.

I also would like to run your new test `test_get_organisation_services_usage_limit_queries_executed` by itself against the main branch to see what the value was before as well. I might try and find 2 mins to do that tomorrow.

Overall though, for the code I've reviewed there isn't anything that seems too controversial in there and it seems very logical. I've left a few comments and generally your changes don't add particular complexity to the code whilst giving big speed improvements.

I haven't run this locally using your dummy data. This is a low traffic page and we should be able to roll this out, see if it is much better and if it is not, we can roll back. I would be more cautious if we were changing something like the notification sending API routes where an accidental degradation in performance could actually cause a cascade of failures and a big incident.

Have a look, see what you think to my comments and then I can potentially review the second commit later in the week if you think that it needs to go in for this PR to be worth it.

Thank you Sam, in particular for getting your head around some of our most complex code and splitting into small commits.

File: app/dao/fact_billing_dao.py [hunk 2]
```diff
@@ -465,24 +465,26 @@ def _fetch_usage_for_service_sms(service_id, year):
...
 
 
-def fetch_billing_data_for_day(process_day, service_id=None, check_permissions=False):
+def fetch_billing_data_for_day(process_day, service_ids=None, check_permissions=False):
     start_date = get_london_midnight_in_utc(process_day)
     end_date = get_london_midnight_in_utc(process_day + timedelta(days=1))
     current_app.logger.info("Populate ft_billing for {} to {}".format(start_date, end_date))
     transit_data = []
-    if not service_id:
+    if service_ids is None:
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/dao/fact_billing_dao.py:484]: Yes, that's why 👍

File: app/commands.py [hunk 3]
```diff
@@ -29,7 +29,7 @@
     set_default_free_allowance_for_service,
 )
 from app.dao.fact_billing_dao import (
-    delete_billing_data_for_service_for_day,
+    delete_billing_data_for_services_for_day,
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/commands.py:34]: Oops, you're right - good spot, thanks! 👀

File: app/celery/reporting_tasks.py [hunk 4]
```diff
@@ -8,7 +8,7 @@
 from app.cronitor import cronitor
 from app.dao.fact_billing_dao import (
     fetch_billing_data_for_day,
-    update_fact_billing,
+    update_billing_facts,
```
[the remark this answers is not in the document]
Comment (samuelhwilliams) [reply app/celery/reporting_tasks.py:11]: My understanding is also that `ft_` stands for `facts`, hence the renames here. Happy to rename to `update_ft_billing` though :)

**Comment by samuelhwilliams** [comment]
> I couldn't easily get my head around the second commit and it got too late for me to start doing it. It probably isn't that crazy but definitely touches some of our more complex code and I would need to do a fair bit of sqlalchemy docs reading. If you want to move this pull request through quicker, I'd be open to moving the second commit to a separate PR or dropping it (assuming that it is not the deal breaker commit that really reduces the bulk of the time). Such complex code of ours is hard to review and I'm not sure that your changes in that commit make it any easier (not that I would necessarily expect them to but if the performance benefits aren't drastic for the commit, I'm OK to not include it for that reason).

It should be possible to drop out the 2nd commit and do it as a separate PR, though we'd be "putting back" a fair number of queries (we'd continue to make 3n rather than 3 queries, where n is the number of services). I think it would be nice to keep in. I've given it another pass and I had some unnecessary separation that made it look like a bigger change than it is. If you look again, hopefully it appears less troublesome now 🤞 

> I also would like to run your new test test_get_organisation_services_usage_limit_queries_executed by itself against the main branch to see what the value was before as well. I might try and find 2 mins to do that tomorrow.

By all means. Hopefully this should be as simple as cherry-picking the commit onto `main` and running/investigating it with `pytest -vvvs -k test_get_organisation_services_usage_limit_queries_executed` or similar. For the 5 services case, I get 26 queries, and for the 10 services case, I get 46 queries.

File: tests/app/organisation/test_rest.py (hunk 1 above)
Comment (samuelhwilliams) [reply tests/app/organisation/test_rest.py]: I've added a bunch more `ft_billing` entries - is that better?

**Comment by samuelhwilliams** [comment]
The failing test is unrelated to this PR (caused by enabling sqlalchemy query recording during tests, and this one single test is triggering some edge case inside sqlalchemy/psycopg2). Looking into ways around it, but shouldn't stop any further reviewing of this PR... 🤦

File: app/celery/reporting_tasks.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (leohemsted) [reply app/celery/reporting_tasks.py]: yeah, it's a fact table (singular) - see https://en.wikipedia.org/wiki/Fact_table. I softly agree with david - given that none of us are data warehousing BI experts, we should just keep referring to it as the table name, so `update_ft_billing` is best

**Review by CrystalPea — COMMENTED** [review]
I know this is merged already, but just had time to look through it, nicely done 👏🏼
